src/System.Private.CoreLib/shared/System/Collections/Generic/NonRandomizedStringEqualityComparer.cs

```diff
@@ -14,21 +14,22 @@ namespace System.Collections.Generic
     // Needs to be public to support binary serialization compatibility
     public sealed class NonRandomizedStringEqualityComparer : EqualityComparer<string?>, ISerializable
     {
-        internal static new IEqualityComparer<string?> Default { get; } = new NonRandomizedStringEqualityComparer();
+        internal static new IEqualityComparer<string?> Default { get; } = new NonRandomizedStringEqualityComparer(EqualityComparer<string?>.Default);
 
-        private NonRandomizedStringEqualityComparer() { }
+        internal static IEqualityComparer<string?> Ordinal { get; } = new NonRandomizedStringEqualityComparer(StringComparer.Ordinal);
+
+        internal IEqualityComparer<string?> BackingComparer { get; }
+
+        private NonRandomizedStringEqualityComparer(IEqualityComparer<string?> backingComparer) => BackingComparer = backingComparer;
 
         // This is used by the serialization engine.
-        private NonRandomizedStringEqualityComparer(SerializationInfo information, StreamingContext context) { }
+        private NonRandomizedStringEqualityComparer(SerializationInfo information, StreamingContext context)
+            : this(EqualityComparer<string?>.Default) { }
 
         public sealed override bool Equals(string? x, string? y) => string.Equals(x, y);
 
         public sealed override int GetHashCode(string? obj) => obj?.GetNonRandomizedHashCode() ?? 0;
 
-        public void GetObjectData(SerializationInfo info, StreamingContext context)
-        {
-            // We are doing this to stay compatible with .NET Framework.
-            info.SetType(typeof(GenericEqualityComparer<string>));
-        }
+        public void GetObjectData(SerializationInfo info, StreamingContext context) => throw new InvalidOperationException();
```

Comment:
Is this change breaking binary serialization compatibility? The comment in the original code explicitly says that this is for binary serialization compatibility with .NET Framework.

Reply:
Cc @viktorhofer

Reply:
It shouldn't break compatibility, because of [the change in `Dictionary.GetObjectData()`](https://github.com/dotnet/coreclr/pull/25883/files/8baf7f952797a243d6541a9dcc9d254349e5060e#diff-c925cc7aff8edb06fbe36ff38cfbfc57R377).

Before, the code directly serialized `_comparer`, which was `NonRandomizedStringEqualityComparer`, so `NonRandomizedStringEqualityComparer` had to make sure that it serializes as `GenericEqualityComparer`.

Now, that code serializes `Comparer` instead, which will be `GenericEqualityComparer`, so `NonRandomizedStringEqualityComparer` does not become relevant during serialization.

I have written a small program to verify that serializing with this code and deserializing on .Net Framework works. [It can be found, along with the output of running it, here.](https://gist.github.com/svick/d001b43f714f78a1dff7f65caaf889c9)

Reply:
This one may be ok. But what about this?

// Required for compatibility with .NET Core 2.0 as we exposed the NonRandomizedStringEqualityComparer inside the serialization blob

Does this still work if you get one of these payloads from .NET Core 2.0 that contain serialized NonRandomizedStringEqualityComparer ?

Reply:
That's why I removed NRSEC serialization, but kept deserialization intact, so this should work.

Here is the output of my test program, showing that this scenario still works (using `<RuntimeFrameworkVersion>2.0.3</RuntimeFrameworkVersion>` for `netcoreapp2.0`):

```
PS C:\code\tmp\comparer-serialization> dotnet run -f netcoreapp2.0 s
comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
Comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
PS C:\code\tmp\comparer-serialization> dotnet run -f netcoreapp2.0 d
comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
Comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
PS C:\code\tmp\comparer-serialization> dotnet run -f net472 d       

Unhandled Exception: System.InvalidCastException: Object must implement IConvertible.
   at System.Convert.ChangeType(Object value, Type conversionType, IFormatProvider provider)
   at System.Runtime.Serialization.SerializationInfo.GetValue(String name, Type type)
   at System.Collections.Generic.Dictionary`2.OnDeserialization(Object sender)
   at System.Runtime.Serialization.ObjectManager.RaiseDeserializationEvent()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream)
   at Program.Main(String[] args) in C:\code\tmp\comparer-serialization\Program.cs:line 38
PS C:\code\tmp\comparer-serialization> dotnet run -f netcoreapp3.0 d
_comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
Comparer: System.Collections.Generic.GenericEqualityComparer`1[System.String]
PS C:\code\tmp\comparer-serialization> .\bin\Debug\netcoreapp3.0\win10-x64\publish\comparer-serialization.exe d
_comparer: System.Collections.Generic.NonRandomizedStringEqualityComparer
Comparer: System.Collections.Generic.GenericEqualityComparer`1[System.String]
```